File: server/src/presentation/services/message.service.ts
import mongoose from "mongoose";
import { ChatModel } from "../../data/mongo/models/chat.model";
import { MessageModel } from "../../data/mongo/models/message.model";
import { CreateMessageDto } from "../../domain/dtos/messages/create-message.dto";
import { MessageEntity } from "../../domain/entities/message.entity";
import { CustomError } from "../../domain/errors/custom.error";
import { UpdateMessageDto } from "../../domain/dtos/messages/update-message.dto";
import { PaginationDto } from "../../domain/dtos/shared/pagination.dto";

export class MessageService {
    constructor() { }

    async getMessagesByChatId(chatId: string, paginationDto: PaginationDto) {
        await this.checkChatStatus(chatId);

        const { page, limit } = paginationDto;

        const id = new mongoose.Types.ObjectId(chatId)

        try {
            const totalPromise = MessageModel.countDocuments({ chat: id });
            const messagesPromise = MessageModel.find({ chat: id })
                .sort({ 'createdAt': 'desc' })
                .skip((page - 1) * limit)
                .limit(limit);

            const [total, messages] = await Promise.all([totalPromise, messagesPromise]);

            const messageEntities = messages.reverse().map(MessageEntity.fromObject);

            return {
                page,
                limit,
                total,
                totalPages: Math.ceil(total / limit),
                next: (page * limit < total) ? `/api/messages/${chatId}?page=${page + 1}&limit=${limit}` : null,
                prev: (page > 1) ? `/api/messages/${chatId}?page=${page - 1}&limit=${limit}` : null,
                messages: messageEntities,
            }
        } catch (error) {
            throw CustomError.internalServer(`${error}`);
        }
    }

    async create(createMessageDto: CreateMessageDto) {
        const { chat: chatId } = createMessageDto;
        await this.checkChatStatus(chatId);

        try {
            const message = new MessageModel(createMessageDto);
            await message.save();
            return MessageEntity.fromObject(message);
        } catch (error) {
            throw CustomError.internalServer(`${error}`);
        }
    }

    async getUnseenMessages(chatIds: string[], userId: string) {
        try {
            const messages = await MessageModel.find({
                isSeen: false,
                chat: { $in: chatIds },
                // sender: { $ne: userId }
            }).select('_id sender chat');

            const unseenMessagesByChat = messages.reduce((acc: { [key: string]: { id: string, sender: string }[] }, message) => {
                const { _id: id, sender, chat } = message;
                const key = chat!.toString();
                if (!(key in acc)) {
                    acc[key] = []
                }
                acc[key].push({ id: id.toString(), sender: sender!.toString() });
                return acc;
            }, {});

            return unseenMessagesByChat;
        } catch (error) {
            console.log(error)
            throw CustomError.internalServer(`${error}`);
        }
    }

    async updateMessagesStatus(updateMessageDto: UpdateMessageDto) {
        const { ids } = updateMessageDto
        try {
            const messages = await MessageModel.updateMany(
                { _id: { $in: ids } },
                { $set: { isSeen: true } }
            );
            return messages;
        } catch (error) {
            throw CustomError.internalServer(`${error}`);
        }
    }

    private async checkChatStatus(chatId: string) {
        const chatExists = await ChatModel.findById(chatId);
        if (!chatExists) throw CustomError.badRequest('Chat does not exists');
    }

}

File: ServerNET/Application/Messages/MessageService.cs
using System;
using Application.Core;
using Application.DTOs;
using Application.DTOs.Messages;
using Domain;
using Microsoft.Extensions.Options;
using MongoDB.Bson;
using MongoDB.Driver;
using Persistence;

namespace Application.Messages;

public class MessageService(AppDbContext dbContext, IOptions<AppDbSettings> settings)
{
    private readonly IMongoCollection<Message> _messagesCollection = dbContext.Database.GetCollection<Message>(settings.Value.MessagesCollectionName);
    public async Task<List<Message>> GetAll()
    {
        return await _messagesCollection.Find(_ => true).ToListAsync();
    }

    public async Task<Result<object>> GetMessagesByChatId(string chatId, PaginationDto paginationDto)
    {
        try
        {
            int page = paginationDto.Page;
            int limit = paginationDto.Limit;

            var filter = Builders<Message>.Filter.Eq(m => m.Chat, chatId);

            var total = await _messagesCollection.CountDocumentsAsync(filter);

            var messages = await _messagesCollection
                .Find(filter)
                .SortByDescending(m => m.CreatedAt)
                .Skip((page - 1) * limit)
                .Limit(limit)
                .ToListAsync();

            messages.Reverse();

            var messagesDto = messages.Select(MessageDTO.FromEntity).ToList();

            var totalPages = (int)Math.Ceiling((double)total / limit);

            var result = new
            {
                page,
                limit,
                total,
                totalPages,
                next = (page * limit < total) ? $"/api/messages/{chatId}?page={page + 1}&limit={limit}" : null,
                prev = (page > 1) ? $"/api/messages/{chatId}?page={page - 1}&limit={limit}" : null,
                messages = messagesDto
            };
            return Result<object>.Success(result);

        }
        catch (Exception ex)
        {
            return Result<object>.Failure($"Internal server error {ex.Message}", 500);
        }
    }

    public async Task<Result<object>> GetUnseenMessages(List<string> chatIds, string userId)
    {
        try
        {
            var filter = Builders<Message>.Filter.And(
                Builders<Message>.Filter.Eq(m => m.IsSeen, false),
                Builders<Message>.Filter.In(m => m.Chat, chatIds)
            // Builders<Message>.Filter.Ne(m => m.Sender, userId)
            );

            var projection = Builders<Message>.Projection
                .Include(m => m.Id)
                .Include(m => m.Sender)
                .Include(m => m.Chat);

            var bsonMessages = await _messagesCollection
                .Find(filter)
                .Project<BsonDocument>(projection)
                .ToListAsync();

            var unseenMessagesByChat = new Dictionary<string, List<UnseenMessageDTO>>();

            foreach (var doc in bsonMessages)
            {
                var chatId = doc["chat"].AsObjectId.ToString();
                var message = new UnseenMessageDTO
                {
                    Id = doc["_id"].AsObjectId.ToString(),
                    Sender = doc["sender"].AsObjectId.ToString()
                };

                if (!unseenMessagesByChat.ContainsKey(chatId))
                    unseenMessagesByChat[chatId] = [];

                unseenMessagesByChat[chatId].Add(message);
            }

            return Result<object>.Success(unseenMessagesByChat);
        }
        catch (Exception ex)
        {
            return Result<object>.Failure($"Internal server error: {ex.Message}", 500);
        }
    }

    public async Task<Result<Message>> Create(CreateMessageDto createMessageDto)
    {
        try
        {
            var message = new Message
            {
                Chat = createMessageDto.Chat,
                Content = createMessageDto.Content,
                Image = createMessageDto.Image,
                IsSeen = createMessageDto.IsSeen,
                Sender = createMessageDto.Sender,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow
            };

            await _messagesCollection.InsertOneAsync(message);

            return Result<Message>.Success(await _messagesCollection.Find(m => m.Id == message.Id).FirstOrDefaultAsync());
        }
        catch (Exception ex)
        {
            return Result<Message>.Failure($"Internal server error: {ex.Message}", 500);
        }
    }
}
